File: examples/python/securesys_client.py
"""
SecureSys Auditor Python Client

A simple, easy-to-use Python client for the SecureSys Auditor API.

Usage:
    from securesys_client import SecureSysClient
    
    client = SecureSysClient(api_url="https://api.securesys.io", api_key="your-key")
    
    # Submit a scan
    scan = client.submit_scan(
        hostname="webserver-01.example.com",
        scan_type="vulnerability",
        findings=[
            {"title": "CVE-2023-1234", "severity": "high", "description": "..."}
        ]
    )
    
    # Get scan status
    status = client.get_scan(scan["id"])
    
    # List all scans
    scans = client.list_scans(page=1, page_size=20)
"""

import json
import logging
from dataclasses import dataclass, field
from datetime import datetime
from typing import Any, Dict, List, Optional, Union
from urllib.parse import urljoin

import requests
from requests.adapters import HTTPAdapter
from urllib3.util.retry import Retry

logger = logging.getLogger(__name__)


@dataclass
class Finding:
    """Represents a security finding."""
    title: str
    severity: str  # critical, high, medium, low, info
    description: str
    recommendation: str = ""
    affected_component: str = ""
    cve_id: Optional[str] = None
    cwe_id: Optional[str] = None
    
    def to_dict(self) -> Dict[str, Any]:
        return {
            "title": self.title,
            "severity": self.severity,
            "description": self.description,
            "recommendation": self.recommendation,
            "affected_component": self.affected_component,
            "cve_id": self.cve_id,
            "cwe_id": self.cwe_id,
        }


@dataclass
class ScanResult:
    """Represents a scan result."""
    id: str
    hostname: str
    scan_type: str
    status: str
    created_at: str
    findings_count: int = 0
    critical_count: int = 0
    high_count: int = 0
    medium_count: int = 0
    low_count: int = 0
    
    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> "ScanResult":
        return cls(
            id=data["id"],
            hostname=data["hostname"],
            scan_type=data["scan_type"],
            status=data["status"],
            created_at=data["created_at"],
            findings_count=data.get("findings_count", 0),
            critical_count=data.get("critical_count", 0),
            high_count=data.get("high_count", 0),
            medium_count=data.get("medium_count", 0),
            low_count=data.get("low_count", 0),
        )


class SecureSysError(Exception):
    """Base exception for SecureSys client errors."""
    def __init__(self, message: str, status_code: Optional[int] = None, response: Optional[Dict] = None):
        super().__init__(message)
        self.status_code = status_code
        self.response = response


class AuthenticationError(SecureSysError):
    """Raised when authentication fails."""
    pass


class RateLimitError(SecureSysError):
    """Raised when rate limit is exceeded."""
    def __init__(self, message: str, retry_after: Optional[int] = None, **kwargs):
        super().__init__(message, **kwargs)
        self.retry_after = retry_after


class ValidationError(SecureSysError):
    """Raised when request validation fails."""
    pass


class SecureSysClient:
    """
    SecureSys Auditor API Client.
    
    Example:
        client = SecureSysClient(
            api_url="https://api.securesys.io",
            api_key="your-api-key"
        )
        
        # Submit a scan
        scan = client.submit_scan(
            hostname="server.example.com",
            scan_type="vulnerability",
            findings=[Finding(title="CVE-2023-1234", severity="high", description="...")]
        )
    """
    
    DEFAULT_TIMEOUT = 30
    DEFAULT_RETRIES = 3
    
    def __init__(
        self,
        api_url: str,
        api_key: str,
        timeout: int = DEFAULT_TIMEOUT,
        retries: int = DEFAULT_RETRIES,
        verify_ssl: bool = True,
    ):
        """
        Initialize the SecureSys client.
        
        Args:
            api_url: Base URL of the SecureSys API
            api_key: API key for authentication
            timeout: Request timeout in seconds
            retries: Number of retries for failed requests
            verify_ssl: Whether to verify SSL certificates
        """
        self.api_url = api_url.rstrip("/")
        self.api_key = api_key
        self.timeout = timeout
        self.verify_ssl = verify_ssl
        
        # Configure session with retries
        self.session = requests.Session()
        retry_strategy = Retry(
            total=retries,
            backoff_factor=1,
            status_forcelist=[429, 500, 502, 503, 504],
            allowed_methods=["GET", "POST", "PUT", "DELETE"],
        )
        adapter = HTTPAdapter(max_retries=retry_strategy)
        self.session.mount("http://", adapter)
        self.session.mount("https://", adapter)
        
        # Set default headers
        self.session.headers.update({
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
            "User-Agent": "SecureSys-Python-Client/1.0",
        })
    
    def _request(
        self,
        method: str,
        endpoint: str,
        data: Optional[Dict] = None,
        params: Optional[Dict] = None,
    ) -> Dict[str, Any]:
        """Make an API request."""
        url = urljoin(self.api_url + "/", endpoint.lstrip("/"))
        
        try:
            response = self.session.request(
                method=method,
                url=url,
                json=data,
                params=params,
                timeout=self.timeout,
                verify=self.verify_ssl,
            )
            
            # Handle errors
            if response.status_code == 401:
                raise AuthenticationError(
                    "Authentication failed. Check your API key.",
                    status_code=401
                )
            elif response.status_code == 429:
                retry_after = response.headers.get("Retry-After")
                raise RateLimitError(
                    "Rate limit exceeded.",
                    status_code=429,
                    retry_after=int(retry_after) if retry_after else None
                )
            elif response.status_code == 400:
                raise ValidationError(
                    f"Validation error: {response.text}",
                    status_code=400,
                    response=response.json() if response.text else None
                )
            elif response.status_code >= 400:
                raise SecureSysError(
                    f"API error: {response.text}",
                    status_code=response.status_code
                )
            
            if response.text:
                return response.json()
            return {}
            
        except requests.exceptions.Timeout:
            raise SecureSysError("Request timed out")
        except requests.exceptions.ConnectionError:
            raise SecureSysError("Connection failed")
    
    # =========================================================================
    # Health & Status
    # =========================================================================
    
    def health_check(self) -> Dict[str, Any]:
        """Check API health status."""
        return self._request("GET", "/api/v1/health/")
    
    # =========================================================================
    # Scans
    # =========================================================================
    
    def submit_scan(
        self,
        hostname: str,
        scan_type: str = "vulnerability",
        findings: Optional[List[Union[Finding, Dict]]] = None,
        agent_version: str = "python-client-1.0",
        metadata: Optional[Dict] = None,
    ) -> ScanResult:
        """
        Submit a new security scan.
        
        Args:
            hostname: Target hostname
            scan_type: Type of scan (vulnerability, compliance, configuration)
            findings: List of findings (Finding objects or dicts)
            agent_version: Version of the scanning agent
            metadata: Additional metadata
        
        Returns:
            ScanResult object with scan details
        """
        # Convert Finding objects to dicts
        findings_data = []
        if findings:
            for f in findings:
                if isinstance(f, Finding):
                    findings_data.append(f.to_dict())
                else:
                    findings_data.append(f)
        
        data = {
            "hostname": hostname,
            "scan_type": scan_type,
            "agent_version": agent_version,
            "started_at": datetime.utcnow().isoformat() + "Z",
            "completed_at": datetime.utcnow().isoformat() + "Z",
            "findings": findings_data,
            "metadata": metadata or {},
        }
        
        result = self._request("POST", "/api/v1/scans/", data=data)
        return ScanResult.from_dict(result)
    
    def get_scan(self, scan_id: str) -> ScanResult:
        """Get details of a specific scan."""
        result = self._request("GET", f"/api/v1/scans/{scan_id}/")
        return ScanResult.from_dict(result)
    
    def list_scans(
        self,
        page: int = 1,
        page_size: int = 20,
        hostname: Optional[str] = None,
        status: Optional[str] = None,
    ) -> Dict[str, Any]:
        """
        List scans with pagination and filtering.
        
        Args:
            page: Page number
            page_size: Number of results per page
            hostname: Filter by hostname
            status: Filter by status
        
        Returns:
            Dict with 'results' list and pagination info
        """
        params = {"page": page, "page_size": page_size}
        if hostname:
            params["hostname"] = hostname
        if status:
            params["status"] = status
        
        return self._request("GET", "/api/v1/scans/", params=params)
    
    def get_scan_findings(self, scan_id: str) -> List[Dict[str, Any]]:
        """Get findings for a specific scan."""
        result = self._request("GET", f"/api/v1/scans/{scan_id}/findings/")
        return result.get("results", [])
    
    # =========================================================================
    # Systems
    # =========================================================================
    
    def list_systems(self, page: int = 1, page_size: int = 20) -> Dict[str, Any]:
        """List all systems."""
        return self._request("GET", "/api/v1/systems/", params={"page": page, "page_size": page_size})
    
    def get_system(self, system_id: str) -> Dict[str, Any]:
        """Get details of a specific system."""
        return self._request("GET", f"/api/v1/systems/{system_id}/")
    
    # =========================================================================
    # Dashboard
    # =========================================================================
    
    def get_dashboard_stats(self) -> Dict[str, Any]:
        """Get dashboard statistics."""
        return self._request("GET", "/api/v1/dashboard/stats/")
    
    def get_severity_breakdown(self) -> Dict[str, Any]:
        """Get severity breakdown for findings."""
        return self._request("GET", "/api/v1/dashboard/severity-breakdown/")
    
    # =========================================================================
    # Webhooks
    # =========================================================================
    
    def list_webhooks(self) -> List[Dict[str, Any]]:
        """List configured webhooks."""
        result = self._request("GET", "/api/v1/webhooks/")
        return result.get("results", [])
    
    def create_webhook(
        self,
        name: str,
        url: str,
        event_types: List[str],
        secret: Optional[str] = None,
    ) -> Dict[str, Any]:
        """
        Create a new webhook endpoint.
        
        Args:
            name: Webhook name
            url: Webhook URL
            event_types: List of events to subscribe to
            secret: HMAC secret for signing payloads
        """
        data = {
            "name": name,
            "url": url,
            "event_types": event_types,
        }
        if secret:
            data["secret"] = secret
        
        return self._request("POST", "/api/v1/webhooks/", data=data)


# Convenience function for quick usage
def create_client(api_url: str, api_key: str, **kwargs) -> SecureSysClient:
    """Create a SecureSys client with default settings."""
    return SecureSysClient(api_url=api_url, api_key=api_key, **kwargs)


if __name__ == "__main__":
    # Example usage
    import os
    
    client = SecureSysClient(
        api_url=os.environ.get("SECURESYS_API_URL", "http://localhost:8000"),
        api_key=os.environ.get("SECURESYS_API_KEY", "test-key"),
    )
    
    # Check health
    print("Health:", client.health_check())
    
    # Submit a scan
    scan = client.submit_scan(
        hostname="example-server.local",
        scan_type="vulnerability",
        findings=[
            Finding(
                title="CVE-2023-1234",
                severity="high",
                description="Example vulnerability",
                recommendation="Update to latest version",
            )
        ],
    )
    print(f"Submitted scan: {scan.id}")

File: examples/typescript/securesys-client.ts
/**
 * SecureSys Auditor TypeScript Client
 *
 * A type-safe client for the SecureSys Auditor API.
 *
 * @example
 * ```typescript
 * import { SecureSysClient, Finding } from './securesys-client';
 *
 * const client = new SecureSysClient({
 *   apiUrl: 'https://api.securesys.io',
 *   apiKey: 'your-api-key'
 * });
 *
 * const scan = await client.submitScan({
 *   hostname: 'webserver-01.example.com',
 *   scanType: 'vulnerability',
 *   findings: [
 *     { title: 'CVE-2023-1234', severity: 'high', description: '...' }
 *   ]
 * });
 * ```
 */

// =============================================================================
// Types
// =============================================================================

export type Severity = 'critical' | 'high' | 'medium' | 'low' | 'info';
export type ScanType = 'vulnerability' | 'compliance' | 'configuration';
export type ScanStatus = 'pending' | 'processing' | 'completed' | 'failed';

export interface Finding {
  title: string;
  severity: Severity;
  description: string;
  recommendation?: string;
  affectedComponent?: string;
  cveId?: string;
  cweId?: string;
}

export interface ScanSubmission {
  hostname: string;
  scanType: ScanType;
  findings: Finding[];
  agentVersion?: string;
  metadata?: Record<string, unknown>;
}

export interface ScanResult {
  id: string;
  hostname: string;
  scanType: ScanType;
  status: ScanStatus;
  createdAt: string;
  completedAt?: string;
  findingsCount: number;
  criticalCount: number;
  highCount: number;
  mediumCount: number;
  lowCount: number;
}

export interface PaginatedResponse<T> {
  count: number;
  next: string | null;
  previous: string | null;
  results: T[];
}

export interface DashboardStats {
  totalSystems: number;
  totalScans: number;
  totalFindings: number;
  criticalFindings: number;
  highFindings: number;
  recentScans: ScanResult[];
}

export interface WebhookEndpoint {
  id: string;
  name: string;
  url: string;
  eventTypes: string[];
  isActive: boolean;
  lastTriggered?: string;
}

export interface ClientConfig {
  apiUrl: string;
  apiKey: string;
  timeout?: number;
  retries?: number;
}

// =============================================================================
// Errors
// =============================================================================

export class SecureSysError extends Error {
  constructor(
    message: string,
    public statusCode?: number,
    public response?: unknown
  ) {
    super(message);
    this.name = 'SecureSysError';
  }
}

export class AuthenticationError extends SecureSysError {
  constructor(message: string = 'Authentication failed') {
    super(message, 401);
    this.name = 'AuthenticationError';
  }
}

export class RateLimitError extends SecureSysError {
  constructor(
    message: string = 'Rate limit exceeded',
    public retryAfter?: number
  ) {
    super(message, 429);
    this.name = 'RateLimitError';
  }
}

export class ValidationError extends SecureSysError {
  constructor(message: string, public errors?: Record<string, string[]>) {
    super(message, 400);
    this.name = 'ValidationError';
  }
}

// =============================================================================
// Client
// =============================================================================

export class SecureSysClient {
  private apiUrl: string;
  private apiKey: string;
  private timeout: number;
  private retries: number;

  constructor(config: ClientConfig) {
    this.apiUrl = config.apiUrl.replace(/\/$/, '');
    this.apiKey = config.apiKey;
    this.timeout = config.timeout ?? 30000;
    this.retries = config.retries ?? 3;
  }

  // ===========================================================================
  // Private Methods
  // ===========================================================================

  private async request<T>(
    method: string,
    endpoint: string,
    options: {
      body?: unknown;
      params?: Record<string, string | number>;
    } = {}
  ): Promise<T> {
    const url = new URL(`${this.apiUrl}${endpoint}`);

    if (options.params) {
      Object.entries(options.params).forEach(([key, value]) => {
        url.searchParams.append(key, String(value));
      });
    }

    const headers: HeadersInit = {
      'Authorization': `Bearer ${this.apiKey}`,
      'Content-Type': 'application/json',
      'User-Agent': 'SecureSys-TypeScript-Client/1.0',
    };

    let lastError: Error | null = null;

    for (let attempt = 0; attempt < this.retries; attempt++) {
      try {
        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), this.timeout);

        const response = await fetch(url.toString(), {
          method,
          headers,
          body: options.body ? JSON.stringify(options.body) : undefined,
          signal: controller.signal,
        });

        clearTimeout(timeoutId);

        if (response.status === 401) {
          throw new AuthenticationError();
        }

        if (response.status === 429) {
          const retryAfter = response.headers.get('Retry-After');
          throw new RateLimitError(
            'Rate limit exceeded',
            retryAfter ? parseInt(retryAfter, 10) : undefined
          );
        }

        if (response.status === 400) {
          const errorData = await response.json();
          throw new ValidationError('Validation failed', errorData);
        }

        if (!response.ok) {
          throw new SecureSysError(
            `API error: ${response.statusText}`,
            response.status
          );
        }

        const text = await response.text();
        return text ? JSON.parse(text) : {} as T;

      } catch (error) {
        lastError = error as Error;

        // Don't retry auth errors or validation errors
        if (
          error instanceof AuthenticationError ||
          error instanceof ValidationError
        ) {
          throw error;
        }

        // Retry on rate limit with backoff
        if (error instanceof RateLimitError && error.retryAfter) {
          await this.sleep(error.retryAfter * 1000);
          continue;
        }

        // Exponential backoff for other errors
        if (attempt < this.retries - 1) {
          await this.sleep(Math.pow(2, attempt) * 1000);
        }
      }
    }

    throw lastError || new SecureSysError('Request failed after retries');
  }

  private sleep(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms));
  }

  private toSnakeCase(obj: Record<string, unknown>): Record<string, unknown> {
    const result: Record<string, unknown> = {};
    for (const [key, value] of Object.entries(obj)) {
      const snakeKey = key.replace(/([A-Z])/g, '_$1').toLowerCase();
      result[snakeKey] = value;
    }
    return result;
  }

  // ===========================================================================
  // Health & Status
  // ===========================================================================

  async healthCheck(): Promise<{ status: string; version: string }> {
    return this.request('GET', '/api/v1/health/');
  }

  // ===========================================================================
  // Scans
  // ===========================================================================

  async submitScan(scan: ScanSubmission): Promise<ScanResult> {
    const now = new Date().toISOString();

    const body = {
      hostname: scan.hostname,
      scan_type: scan.scanType,
      agent_version: scan.agentVersion || 'typescript-client-1.0',
      started_at: now,
      completed_at: now,
      findings: scan.findings.map(f => ({
        title: f.title,
        severity: f.severity,
        description: f.description,
        recommendation: f.recommendation || '',
        affected_component: f.affectedComponent || '',
        cve_id: f.cveId,
        cwe_id: f.cweId,
      })),
      metadata: scan.metadata || {},
    };

    const result = await this.request<Record<string, unknown>>('POST', '/api/v1/scans/', { body });
    return this.parseScanResult(result);
  }

  async getScan(scanId: string): Promise<ScanResult> {
    const result = await this.request<Record<string, unknown>>('GET', `/api/v1/scans/${scanId}/`);
    return this.parseScanResult(result);
  }

  async listScans(options?: {
    page?: number;
    pageSize?: number;
    hostname?: string;
    status?: ScanStatus;
  }): Promise<PaginatedResponse<ScanResult>> {
    const params: Record<string, string | number> = {
      page: options?.page ?? 1,
      page_size: options?.pageSize ?? 20,
    };

    if (options?.hostname) params.hostname = options.hostname;
    if (options?.status) params.status = options.status;

    const result = await this.request<{
      count: number;
      next: string | null;
      previous: string | null;
      results: Record<string, unknown>[];
    }>('GET', '/api/v1/scans/', { params });

    return {
      ...result,
      results: result.results.map(r => this.parseScanResult(r)),
    };
  }

  async getScanFindings(scanId: string): Promise<Finding[]> {
    const result = await this.request<{ results: Finding[] }>(
      'GET',
      `/api/v1/scans/${scanId}/findings/`
    );
    return result.results;
  }

  // ===========================================================================
  // Systems
  // ===========================================================================

  async listSystems(options?: {
    page?: number;
    pageSize?: number;
  }): Promise<PaginatedResponse<{ id: string; hostname: string; lastScan?: string }>> {
    return this.request('GET', '/api/v1/systems/', {
      params: {
        page: options?.page ?? 1,
        page_size: options?.pageSize ?? 20,
      },
    });
  }

  // ===========================================================================
  // Dashboard
  // ===========================================================================

  async getDashboardStats(): Promise<DashboardStats> {
    return this.request('GET', '/api/v1/dashboard/stats/');
  }

  async getSeverityBreakdown(): Promise<Record<Severity, number>> {
    return this.request('GET', '/api/v1/dashboard/severity-breakdown/');
  }

  // ===========================================================================
  // Webhooks
  // ===========================================================================

  async listWebhooks(): Promise<WebhookEndpoint[]> {
    const result = await this.request<{ results: WebhookEndpoint[] }>(
      'GET',
      '/api/v1/webhooks/'
    );
    return result.results;
  }

  async createWebhook(webhook: {
    name: string;
    url: string;
    eventTypes: string[];
    secret?: string;
  }): Promise<WebhookEndpoint> {
    return this.request('POST', '/api/v1/webhooks/', {
      body: {
        name: webhook.name,
        url: webhook.url,
        event_types: webhook.eventTypes,
        secret: webhook.secret,
      },
    });
  }

  // ===========================================================================
  // Helpers
  // ===========================================================================

  private parseScanResult(data: Record<string, unknown>): ScanResult {
    return {
      id: data.id as string,
      hostname: data.hostname as string,
      scanType: data.scan_type as ScanType,
      status: data.status as ScanStatus,
      createdAt: data.created_at as string,
      completedAt: data.completed_at as string | undefined,
      findingsCount: (data.findings_count as number) ?? 0,
      criticalCount: (data.critical_count as number) ?? 0,
      highCount: (data.high_count as number) ?? 0,
      mediumCount: (data.medium_count as number) ?? 0,
      lowCount: (data.low_count as number) ?? 0,
    };
  }
}

// =============================================================================
// Factory Function
// =============================================================================

export function createClient(config: ClientConfig): SecureSysClient {
  return new SecureSysClient(config);
}

// =============================================================================
// Example Usage
// =============================================================================

/*
const client = new SecureSysClient({
  apiUrl: 'https://api.securesys.io',
  apiKey: 'your-api-key',
});

// Submit a scan
const scan = await client.submitScan({
  hostname: 'webserver-01.example.com',
  scanType: 'vulnerability',
  findings: [
    {
      title: 'CVE-2023-1234',
      severity: 'high',
      description: 'A vulnerability was found',
      recommendation: 'Update to the latest version',
    },
  ],
});

console.log(`Submitted scan: ${scan.id}`);

// List recent scans
const scans = await client.listScans({ pageSize: 10 });
console.log(`Found ${scans.count} scans`);
*/
